db/migrate/20180109102659_add_unique_indexes_to_containers_tables.rb

```diff
@@ -0,0 +1,87 @@
+  def change
+    # Just having :ems_id & :ems_ref
+    add_index :container_builds,         %i(ems_id ems_ref), :unique => true
+    add_index :container_build_pods,     %i(ems_id ems_ref), :unique => true
+    add_index :container_groups,         %i(ems_id ems_ref), :unique => true
+    add_index :container_limits,         %i(ems_id ems_ref), :unique => true
+    add_index :container_nodes,          %i(ems_id ems_ref), :unique => true
+    add_index :container_projects,       %i(ems_id ems_ref), :unique => true
+    add_index :container_quotas,         %i(ems_id ems_ref), :unique => true
+    add_index :container_replicators,    %i(ems_id ems_ref), :unique => true
+    add_index :container_routes,         %i(ems_id ems_ref), :unique => true
+    add_index :container_services,       %i(ems_id ems_ref), :unique => true
+    add_index :container_templates,      %i(ems_id ems_ref), :unique => true
+    add_index :containers,               %i(ems_id ems_ref), :unique => true
+    add_index :persistent_volume_claims, %i(ems_id ems_ref), :unique => true
+
+    # Having :ems_id but not ems_ref
+    add_index :container_images,
+              %i(ems_id image_ref),
+              :unique => true,
+              :name   => "index_container_images_unique_multi_column"
+    add_index :container_image_registries,
+              %i(ems_id host port),
+              :unique => true
+
+    # Nested tables, not having :ems_id and the foreign_key is a part of the unique index
+    add_index :container_conditions,
+              %i(container_entity_id container_entity_type name),
+              :unique => true,
+              :name   => "index_container_conditions_unique_multi_column"
+    add_index :security_contexts,
+              %i(resource_id resource_type),
+              :unique => true,
+              :name   => "index_security_contexts_unique_multi_column"
+    add_index :taggings,
+              %i(taggable_id taggable_type tag_id),
+              :unique => true,
+              :name   => "index_taggings_unique_multi_column"
+    add_index :computer_systems,
+              %i(managed_entity_id managed_entity_type),
+              :unique => true,
+              :name   => "index_computer_systems_unique_multi_column"
+    add_index :container_env_vars,
+              %i(container_id name value field_path),
+              :unique => true,
+              :name   => "index_container_env_vars_unique_multi_column"
+    add_index :container_limit_items,
+              %i(container_limit_id resource item_type),
+              :unique => true,
+              :name   => "index_container_limit_items_unique_multi_column"
+    add_index :container_port_configs,
+              %i(container_id ems_ref),
+              :unique => true,
+              :name   => "index_container_port_configs_unique_multi_column"
+    add_index :container_quota_items,
+              %i(container_quota_id resource),
```

Comment:
We're going to start keeping history via archiving in this table, and unlike usual soft delete when the ems_ref or other unique id went away, these will collide on same container_quota_id resource.
- Should the index be limited to active (delete_on is null) items?
  But you didn't do this in any other table, and I think it's good to keep uniformity.
- Can you add `quota_desired quota_enforced quota_observed` here?

Reply:
ok, adding ``` :where  => "deleted_on IS NULL",```, for quota items, then with the AR scope, the refresh should leave the archived items be

Reply:
hm, the unique index should then also be ``` [:resource, :quota_desired, :quota_enforced, :quota_observed]``` and the same for manager ref definition

Reply:
@cben is any of ```[:resource, :quota_desired, :quota_enforced, :quota_observed]``` nullable? (we should answer the question and add NOT NULL constraints really, to bring some order to the DB)

Reply:
quota_enforced & quota_observed come from the `status` part => I think currently, quickly after creating the quota, they might be null.  I need to add a test based on transient states recorded with a watch.

We can modify parser, though not immediately sure how.  Putting enforced=0 or observed=0 when this quota is not enforced yet is semantically wrong...  Perhaps we may simply discard quotas that have no status yet.

Reply:
ok, if we can discard them, it would be ok

In the other case ,we would need several partial unique indexes. And would prepare smart batching, so the could would still work I think. But having set of NOT NULL unique indexes would be most effective